File: src/interface/src/robot_interface.cpp
#include <cstdio>
#include <iostream>
#include <unordered_map>
#include "robot_interface.h"
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include <publisher.hpp>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"


using namespace std;

/*
 * Constructors/Destructors
 */
RobotInterface::RobotInterface(Node rosNode) {
    self.selfNode = rosNode;
    cout << "GOT TO INIT" << endl;
    
    // Drive base
    publisher_ = this->create_publisher<float>("front_left_wheel_motor_velocity_from_interface", 10);
    robotPublishers['front_left_wheel_motor'] = publisher_;
    publisher_ = this->create_publisher<float>("front_right_wheel_motor_velocity_from_interface", 10);
    robotPublishers['front_right_wheel_motor'] = publisher_;
    publisher_ = this->create_publisher<float>("mid_left_wheel_motor_velocity_from_interface", 10);
    robotPublishers['mid_left_wheel_motor'] = publisher_;
    publisher_ = this->create_publisher<float>("mid_right_wheel_motor_velocity_from_interface", 10);
    robotPublishers['mid_right_wheel_motor'] = publisher_;
    publisher_ = this->create_publisher<float>("rear_left_wheel_motor_velocity_from_interface", 10);
    robotPublishers['rear_left_wheel_motor'] = publisher_;
    publisher_ = this->create_publisher<float>("rear_right_wheel_motor_velocity_from_interface", 10);
    robotPublishers['rear_right_wheel_motor'] = publisher_;

    // Arm
    publisher_ = this->create_publisher<float>("arm_turntable_motor_position_from_interface", 10);
    robotPublishers['arm_turntable_motor'] = publisher_;
    publisher_ = this->create_publisher<float>("arm_shoulder_motor_position_from_interface", 10);
    robotPublishers['arm_shoulder_motor'] = publisher_;
    publisher_ = this->create_publisher<float>("arm_elbow_motor_position_from_interface", 10);
    robotPublishers['arm_elbow_motor'] = publisher_;
    publisher_ = this->create_publisher<float>("arm_forearm_motor_position_from_interface", 10);
    robotPublishers['arm_forearm_motor'] = publisher_;
    publisher_ = this->create_publisher<float>("arm_wrist_motor_position_from_interface", 10);
    robotPublishers['arm_wrist_motor'] = publisher_;
    publisher_ = this->create_publisher<float>("arm_hand_motor_position_from_interface", 10);
    robotPublishers['arm_hand_motor'] = publisher_;
    publisher_ = this->create_publisher<float>("arm_fingers_motor_position_from_interface", 10);
    robotPublishers['arm_fingers_motor'] = publisher_;

    // Antenna
    publisher_ = this->create_publisher<float>("antenna_motor_position_from_interface", 10);
    robotPublishers['antenna_motor'] = publisher_;
    publisher_ = this->create_publisher<float>("antenna_turntable_motor_position_from_interface", 10);
    robotPublishers['antenna_turntable_motor'] = publisher_;
}

RobotInterface::~RobotInterface() {

}

RobotInterface::RobotInterface(const RobotInterface &) = delete;
RobotInterface &RobotInterface::operator=(const RobotInterface &) = delete;
RobotInterface::RobotInterface(RobotInterface &&) = delete;
RobotInterface &RobotInterface::operator=(RobotInterface &&) = delete;


/*
 * Movement
 */

// General movement including all wheels
void RobotInterface::moveForward(const double amount) {
    
}

void RobotInterface::moveBackward(const double amount) {

}

void RobotInterface::turnLeft(const double amount) {

}

void RobotInterface::turnRight(const double amount) {

}

// Left front wheel
void RobotInterface::front_left_wheel_motor(const double amount) {
    cout << "Testing front_left_wheel_motor" << endl;

    //Converting from input "meters per second" to output "revolutions per second"
    double revolutionsOutput = 1.76838724851 * amount;

    publisher_ = robotPublishers['front_left_wheel_motor']
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing front_left_wheel_motor: " + message.data << endl;
}

// Right front wheel
void RobotInterface::front_right_wheel_motor(const double amount) {
    cout << "Testing front_right_wheel_motor" << endl;

    //Converting from input "meters per second" to output "revolutions per second"
    double revolutionsOutput = 1.76838724851 * amount;

    publisher_ = robotPublishers['front_right_wheel_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing front_right_wheel_motor: " + message.data << endl;
}

// Left middle wheel
void RobotInterface::mid_left_wheel_motor(const double amount) {
    cout << "Testing mid_left_wheel_motor" << endl;

    //Converting from input "meters per second" to output "revolutions per second"
    double revolutionsOutput = 1.76838724851 * amount;

    publisher_ = robotPublishers['mid_left_wheel_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing mid_left_wheel_motor: " + message.data << endl;
}

// Right middle wheel
void RobotInterface::mid_right_wheel_motor(const double amount) {
    cout << "Testing mid_right_wheel_motor" << endl;

    //Converting from input "meters per second" to output "revolutions per second"
    double revolutionsOutput = 1.76838724851 * amount;

    publisher_ = robotPublishers['mid_right_wheel_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing mid_right_wheel_motor: " + message.data << endl;
}

// Left back wheel
void RobotInterface::rear_left_wheel_motor(const double amount) {
    cout << "Testing rear_left_wheel_motor" << endl;

    //Converting from input "meters per second" to output "revolutions per second"
    double revolutionsOutput = 1.76838724851 * amount;

    publisher_ = robotPublishers['rear_left_wheel_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing rear_left_wheel_motor: " + message.data << endl;
}

// Right back wheel
void RobotInterface::rear_right_wheel_motor(const double amount) {
    cout << "Testing rear_right_wheel_motor" << endl;

    //Converting from input "meters per second" to output "revolutions per second"
    double revolutionsOutput = 1.76838724851 * amount;

    publisher_ = robotPublishers['rear_right_wheel_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing rear_right_wheel_motor: " + message.data << endl;
}

/*
 * Arm movement/rotation
 */

// Arm turntable
void RobotInterface::arm_turntable_motor(const double amount) {
    cout << "Testing arm_turntable_motor" << endl;

    //Converting from input "degrees" to output "revolutions"
    double revolutionsOutput = amount / 360;

    publisher_ = robotPublishers['arm_turntable_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing arm_turntable_motor: " + message.data << endl;
}

// Arm 1 is closest to base and arm 3 is furthest from base

// Arm 1 movement
void RobotInterface::arm_shoulder_motor(const double amount) {
    cout << "Testing arm_shoulder_motor" << endl;

    //Converting from input "degrees" to output "revolutions"
    double revolutionsOutput = amount / 360;

    publisher_ = robotPublishers['arm_shoulder_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing arm_shoulder_motor: " + message.data << endl;
}

// Arm 2 movement
void RobotInterface::arm_elbow_motor(const double amount) {
    cout << "Testing arm_elbow_motor" << endl;

    //Converting from input "degrees" to output "revolutions"
    double revolutionsOutput = amount / 360;

    publisher_ = robotPublishers['arm_elbow_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing arm_elbow_motor: " + message.data << endl;
}

// Arm 3 movement
void RobotInterface::arm_forearm_motor(const double amount) {
    cout << "Testing arm_forearm_motor" << endl;

    //Converting from input "degrees" to output "revolutions"
    double revolutionsOutput = amount / 360;

    publisher_ = robotPublishers['arm_forearm_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing arm_forearm_motor: " + message.data << endl;
}

// Arm wrist movement
void RobotInterface::arm_wrist_motor(const double amount) {
    cout << "Testing arm_wrist_motor" << endl;

    //Converting from input "degrees" to output "revolutions"
    double revolutionsOutput = amount / 360;

    publisher_ = robotPublishers['arm_wrist_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing arm_wrist_motor: " + message.data << endl;
}

// Arm hand movement
void RobotInterface::arm_hand_motor(const double amount) {
    cout << "Testing arm_hand_motor" << endl;

    //Converting from input "degrees" to output "revolutions"
    double revolutionsOutput = amount / 360;

    publisher_ = robotPublishers['arm_hand_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing arm_hand_motor: " + message.data << endl;
}

// Arm fingers movement
void RobotInterface::arm_fingers_motor(const double amount) {
    cout << "Testing arm_fingers_motor" << endl;

    //Converting from input "degrees" to output "revolutions"
    double revolutionsOutput = amount / 360;

    publisher_ = robotPublishers['arm_fingers_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing arm_fingers_motor: " + message.data << endl;
}

// antenna_motor movement
void RobotInterface::antenna_motor_extend() {
    cout << "Testing antenna_motor" << endl;

    // Output is in "revolutions"
    doublerevolutionsOutput = 1; // we don't know the actual position it should be. this is a placeholder

    publisher_ = robotPublishers['antenna_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing antenna_motor: " + message.data << endl;
}

void RobotInterface::antenna_motor_retract() {
    cout << "Testing antenna_motor" << endl;

    // Output is in "revolutions"
    doublerevolutionsOutput = 0; // we don't know the actual position it should be. this is a placeholder

    publisher_ = robotPublishers['antenna_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing antenna_motor: " + message.data << endl;
}

void RobotInterface::antenna_turntable_motor(const double amount) {
    cout << "Testing antenna_turntable_motor" << endl;

    // Converting from input "degrees" to output "revolutions"
    double revolutionsOutput = amount / 360

    publisher_ = robotPublishers['antenna_turntable_motor'];
    auto message = std_msgs::msg::String();
    message.data = std::to_string(revolutionsOutput);
    RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
    publisher_->publish(message);
    cout << "Publishing antenna_turntable_motor: " + message.data << endl;
}

File: src/interface/robot_interface.py
# General interface for robot funcitonality

from rclpy.node import Node
from rclpy.publisher import Publisher
from std_msgs.msg import Float32, String

from utility.moteus_data_in_json_helper import MoteusDataInJsonHelper

VELOCITY_CONVERSION = 1
POSITION_CONVERSION = 1 / (6.28)  # Radians to position ticks
REV_TO_RADIANS = 6.28  # One Revolution = 2pi


robot_publishers: dict[str, Publisher] = {}


class RobotInterface:
    #
    # Constructors/Destructors
    #

    def __init__(self, ros_node: Node) -> None:
        self._ros_node = ros_node

        # Drive base
        publisher = self._ros_node.create_publisher(
            String, "front_left_drive_motor_from_interface", 10
        )
        robot_publishers["front_left_drive_motor"] = publisher
        publisher = self._ros_node.create_publisher(
            String, "front_right_drive_motor_from_interface", 10
        )
        robot_publishers["front_right_drive_motor"] = publisher
        publisher = self._ros_node.create_publisher(
            String, "mid_left_drive_motor_from_interface", 10
        )
        robot_publishers["mid_left_drive_motor"] = publisher
        publisher = self._ros_node.create_publisher(
            String, "mid_right_drive_motor_from_interface", 10
        )
        robot_publishers["mid_right_drive_motor"] = publisher
        publisher = self._ros_node.create_publisher(
            String, "rear_left_drive_motor_from_interface", 10
        )
        robot_publishers["rear_left_drive_motor"] = publisher
        publisher = self._ros_node.create_publisher(
            String, "rear_right_drive_motor_from_interface", 10
        )
        robot_publishers["rear_right_drive_motor"] = publisher

        robot_publishers["arm_fingers_motor"] = publisher

        # Antenna
        publisher = self._ros_node.create_publisher(
            Float32, "antenna_motor_position_from_interface", 10
        )
        robot_publishers["antenna_motor"] = publisher
        publisher = self._ros_node.create_publisher(
            Float32, "antenna_turntable_motor_position_from_interface", 10
        )
        robot_publishers["antenna_turntable_motor"] = publisher

        # Arm
        publisher = self._ros_node.create_publisher(
            String, "arm_turntable_motor_from_interface", 10
        )
        robot_publishers["arm_turntable_motor"] = publisher
        publisher = self._ros_node.create_publisher(String, "arm_shoulder_motor_from_interface", 10)
        robot_publishers["arm_shoulder_motor"] = publisher
        publisher = self._ros_node.create_publisher(String, "arm_elbow_motor_from_interface", 10)
        robot_publishers["arm_elbow_motor"] = publisher
        publisher = self._ros_node.create_publisher(
            String, "arm_left_wrist_motor_from_interface", 10
        )
        robot_publishers["arm_left_wrist_motor"] = publisher
        publisher = self._ros_node.create_publisher(
            String, "arm_right_wrist_motor_from_interface", 10
        )
        robot_publishers["arm_right_wrist_motor"] = publisher

    def velocityConversion(self, amount: float) -> float:
        return amount * VELOCITY_CONVERSION

    # Converting from input "radians" to output "revolutions"
    def positionConversion(self, amount: float) -> float:
        return amount * POSITION_CONVERSION

    # General movement including all wheels
    def moveForward(self, amount: float) -> None:
        print()

    def moveBackward(self, amount: float) -> None:
        print()

    def turnLeft(self, amount: float) -> None:
        print()

    def turnRight(self, amount: float) -> None:
        print()

    # ------ DRIVEBASE ------
    # Left front wheel - MEASURED IN REVOLUTIONS PER SECOND
    def frontLeftDriveMotor(self, amount: float) -> None:
        self._sendDriveMotor("front_left_drive_motor", radians_per_second=amount)

    # Right front wheel
    def frontRightDriverMotor(self, amount: float) -> None:
        self._sendDriveMotor("front_right_drive_motor", radians_per_second=amount)

    # Left middle wheel
    def midLeftDriveMotor(self, amount: float) -> None:
        self._sendDriveMotor("mid_left_drive_motor", radians_per_second=amount)

    # Right middle wheel
    def midRightDriveMotor(self, amount: float) -> None:
        self._sendDriveMotor("mid_right_drive_motor", radians_per_second=amount)

    # Left back wheel
    def rearLeftDriveMotor(self, amount: float) -> None:
        self._sendDriveMotor("rear_left_drive_motor", radians_per_second=amount)

    # Right back wheel
    def rearRightDriveMotor(self, amount: float) -> None:
        self._sendDriveMotor("rear_right_drive_motor", radians_per_second=amount)

    def _sendDriveMotor(self, pub_name: str, radians_per_second: float = 0.0) -> None:
        publisher = robot_publishers[pub_name]

        str_msg = String()
        json_helper = MoteusDataInJsonHelper()
        json_helper.velocity = radians_per_second
        json_helper.set_stop = False

        str_msg.data = json_helper.buildJSONString()
        publisher.publish(str_msg)

    # ------ ARM ------

    def armTurntableVelocity(self, target_vel: float) -> None:
        self._sendArmVelocity("arm_turntable_motor", target_vel)

    def shoulderVelocity(self, target_vel: float) -> None:
        self._sendArmVelocity("arm_shoulder_motor", target_vel)

    def elbowVelocity(self, target_vel: float) -> None:
        self._sendArmVelocity("arm_elbow_motor", target_vel)

    def leftWriteVelocity(self, target_vel: float) -> None:
        self._sendArmVelocity("arm_left_wrist_motor", target_vel)

    def rightWristVelocity(self, target_vel: float) -> None:
        self._sendArmVelocity("arm_right_wrist_motor", target_vel)

    # For now, I am sending rev/sec
    def _sendArmVelocity(self, pub_name: str, velocity: float = 0.0) -> None:
        publisher = robot_publishers[pub_name]

        str_msg = String()
        json_helper = MoteusDataInJsonHelper()
        json_helper.velocity = velocity
        json_helper.set_stop = False

        str_msg.data = json_helper.buildJSONString()
        publisher.publish(str_msg)

    def disableArmTurntableMotor(self) -> None:
        self._disableMotor("arm_turntable_motor")

    def disableArmShoulderMotor(self) -> None:
        self._disableMotor("arm_shoulder_motor")

    def disableArmElbowMotor(self) -> None:
        self._disableMotor("arm_elbow_motor")

    def disableArmLeftWristMotor(self) -> None:
        self._disableMotor("arm_left_wrist_motor")

    def disableArmRightWristMotor(self) -> None:
        self._disableMotor("arm_right_wrist_motor")

    def _disableMotor(self, pub_name: str) -> None:
        publisher = robot_publishers[pub_name]

        str_msg = String()
        json_helper = MoteusDataInJsonHelper()
        json_helper.velocity = 0  # for extra security
        json_helper.set_stop = True

        str_msg.data = json_helper.buildJSONString()
        publisher.publish(str_msg)

    # Antenna movement
    def antennaMotorExtend(self) -> None:
        # Output is in "revolutions"
        revolutions_output = (
            1  # we don't know the actual position it should be. this is a placeholder
        )
        publisher = robot_publishers["antenna_motor"]
        str_msg = Float32()
        str_msg.data = revolutions_output
        publisher.publish(str_msg)

    def antennaMotorRetract(self) -> None:
        # Output is in "revolutions"
        revolutions_output = (
            0  # we don't know the actual position it should be. this is a placeholder
        )
        publisher = robot_publishers["antenna_motor"]
        str_msg = Float32()
        str_msg.data = revolutions_output
        publisher.publish(str_msg)

    def antennaTurntableMotor(self, amount: float) -> None:
        revolutions_output = self.positionConversion(amount)
        publisher = robot_publishers["antenna_turntable_motor"]
        str_msg = Float32()
        str_msg.data = revolutions_output
        publisher.publish(str_msg)
